tests/tamr_client/datasets/test_record.py

```diff
@@ -0,0 +1,139 @@
+from functools import partial
+import json
+from typing import Dict
+
+import pytest
+import responses
+
+import tamr_client as tc
+import tests.tamr_client.utils as utils
+
+
+@responses.activate
+def test_update():
+    s = utils.session()
+    dataset = utils.dataset()
+
+    url = tc.URL(path="datasets/1:updateRecords")
+    updates = records_to_updates(_records_json)
+    snoop: Dict = {}
+    responses.add_callback(
+        responses.POST, url.__str__(), partial(create_callback, snoop=snoop, status=200)
+    )
+
+    response = tc.record._update(s, dataset, updates)
+    assert response == _response_json
+    assert snoop["payload"] == stringify(updates)
+
+
+@responses.activate
+def test_upsert():
+    s = utils.session()
+    dataset = utils.dataset()
+
+    url = tc.URL(path="datasets/1:updateRecords")
+    updates = records_to_updates(_records_json)
+    snoop: Dict = {}
+    responses.add_callback(
+        responses.POST, url.__str__(), partial(create_callback, snoop=snoop, status=200)
+    )
+
+    response = tc.record.upsert(
+        s, dataset, _records_json, primary_key_name="primary_key"
+    )
+    assert response == _response_json
+    assert snoop["payload"] == stringify(updates)
+
+
+@responses.activate
+def test_upsert_primary_key_not_found():
+    s = utils.session()
+    dataset = utils.dataset()
+
+    with pytest.raises(tc.record.PrimaryKeyNotFound):
+        tc.record.upsert(
+            s, dataset, _records_json, primary_key_name="wrong_primary_key"
+        )
+
+
+@responses.activate
+def test_delete_by_ids():
+    s = utils.session()
+    dataset = utils.dataset()
+
+    url = tc.URL(path="datasets/1:updateRecords")
+    deletes = records_to_deletes(_records_json)
+    snoop: Dict = {}
+    responses.add_callback(
+        responses.POST, url.__str__(), partial(create_callback, snoop=snoop, status=200)
+    )
+
+    ids = (r["primary_key"] for r in _records_json)
+
+    response = tc.record._delete_by_id(s, dataset, ids)
+    assert response == _response_json
+    assert snoop["payload"] == stringify(deletes)
+
+
+@responses.activate
+def test_delete():
+    s = utils.session()
+    dataset = utils.dataset()
+
+    url = tc.URL(path="datasets/1:updateRecords")
+    deletes = records_to_deletes(_records_json)
+    snoop: Dict = {}
+    responses.add_callback(
+        responses.POST, url.__str__(), partial(create_callback, snoop=snoop, status=200)
+    )
+
+    response = tc.record.delete(
+        s, dataset, _records_json, primary_key_name="primary_key"
+    )
+    assert response == _response_json
+    assert snoop["payload"] == stringify(deletes)
+
+
+@responses.activate
+def test_delete_primary_key_not_found():
+    s = utils.session()
+    dataset = utils.dataset()
+
+    with pytest.raises(tc.record.PrimaryKeyNotFound):
+        tc.record.delete(
+            s, dataset, _records_json, primary_key_name="wrong_primary_key"
+        )
+
+
+def create_callback(request, snoop, status):
```

Comment:
Nice abstraction! Name-wise, what do you think of something like:

```suggestion
def capture_payload(request, snoop, status):
```

or something more descriptive than `create_callback`?

Might be nice to also include a docstring for this function that points back to an example from the `responses` docs

Reply:
Agreed on the function name.  Does this seem like something we will want to use in other tests going forward, and if so, should it be located in a different file and then imported?

Also, I'm not sure how to go about pointing to the `responses` doc.

Reply:
Let's keep in this file for now, but if we see that we need a similar/same function in other test files, we can *then* factor it out until a more reusable place.

As for docstring linking to `responses` docs, how about:

```python
def _capture_payload(request, snoop, status):
    """Capture request body within `snoop` so we can inspect that the request body is constructed correctly (e.g. for streaming requests).

    See https://github.com/getsentry/responses#dynamic-responses
    """
```

Reply:
Hmm, just realized that we do already use this across `test_dataframe` and `test_record` (along with the other helper functions).

We should probably move those to a shared location. `tests.utils` is fine for now

Reply:
Should the variables, e.g. `_records_json` stay in test files themselves?

Reply:
Ideally, we could get some data that looks like *real* records rather than just `{"primary_key": 1}` and then we could store the real data as a `.json` (or `.ndjson` for multiple records).

But I'm ok with keeping as is with minor duplication for the data since I could imagine wanting to test these interfaces against different looking data in the future.